### ISSUE TITLE
Database failover - closed connections returned from pool

### ISSUE THREAD

**Issue by janfh** [brettwooldridge/HikariCP#1296]
#### Environment
```
HikariCP version: 3.2.0
JDK version     : 1.8.0_111
Database        : SQL Server 2017
Driver version  : 7.0.0
```
We have recently moved to a new SQL Server 2017 Always On Availability Groups cluster. The application is based on Spring 5, Hibernate and HikariCP. Everything is working fine, but when we did failover testing the application logged exceptions when the failover happened. The application recovered shortly afterwards (̃~1 second). I suspect that connections already in the pool were not tested on borrow causing the exception due to closed/reset connection to the database. New requests for connections afterwards apparently connected to the new master correctly. 

`
12-Dec-2018 21:51:22.435 SEVERE [https-jsse-nio-443-exec-6] org.apache.catalina.core.StandardWrapperValve.invoke Servlet.service() for servlet [dispatcher] in context with path [] threw exception [Request processing failed; nested exception is org.springframework.transaction.CannotCreateTransactionException: Could not open JPA EntityManager for transaction; nested exception is org.hibernate.TransactionException: JDBC begin transaction failed: ] with root cause
 java.net.SocketException: Connection reset
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:115)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
	at com.microsoft.sqlserver.jdbc.TDSChannel.write(IOBuffer.java:1988)
	at com.microsoft.sqlserver.jdbc.TDSWriter.flush(IOBuffer.java:4096)
	at com.microsoft.sqlserver.jdbc.TDSWriter.writePacket(IOBuffer.java:3998)
	at com.microsoft.sqlserver.jdbc.TDSWriter.endMessage(IOBuffer.java:3128)
	at com.microsoft.sqlserver.jdbc.TDSCommand.startResponse(IOBuffer.java:7602)
	at com.microsoft.sqlserver.jdbc.TDSCommand.startResponse(IOBuffer.java:7590)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection$1ConnectionCommand.doExecute(SQLServerConnection.java:2907)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7240)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:2869)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.connectionCommand(SQLServerConnection.java:2912)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.setAutoCommit(SQLServerConnection.java:3045)
	at com.zaxxer.hikari.pool.ProxyConnection.setAutoCommit(ProxyConnection.java:388)
	at com.zaxxer.hikari.pool.HikariProxyConnection.setAutoCommit(HikariProxyConnection.java)
	at org.hibernate.resource.jdbc.internal.AbstractLogicalConnectionImplementor.begin(AbstractLogicalConnectionImplementor.java:67)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.begin(LogicalConnectionManagedImpl.java:263)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.begin(JdbcResourceLocalTransactionCoordinatorImpl.java:236)
	at org.hibernate.engine.transaction.internal.TransactionImpl.begin(TransactionImpl.java:80)
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.beginTransaction(HibernateJpaDialect.java:183)
	at org.springframework.orm.jpa.JpaTransactionManager.doBegin(JpaTransactionManager.java:401)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:378)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.createTransactionIfNecessary(TransactionAspectSupport.java:474)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:289)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at xxx.xxx.service.xxxxx.xxxxService$$EnhancerBySpringCGLIB$$d1eb44bc.findByXXXX(<generated>)

`

According to the documentation it looks like connections should be tested before they are given to the application.

Is there any configuration needed perhaps in order to get the connections tested on borrow?

**Comment by brettwooldridge** [comment]
HikariCP, by default, has a performance optimization such that connections used within the past 500ms are considered valid (as they were *just* used) — an therefore validation is skipped.  This is called the [bypass window](
https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/HikariPool.java#L81).

Users *think* they want to validate on every borrow, by this has two pitfalls.  First, doing so will absolutely crush your performance, likely cutting it nearly in half; as well as increasing the server load significantly.

Second, it actually doesn’t provide any additional guarantees; it is a classic race condition.  Just as the connection that *was* valid 500ms before, is not valid now — because the database failed *during* those 500ms — as soon as a theoretical validation *would* complete it is possible for the database to fail before the user’s query executes, and encounter a similar error.

Basically, even validating every connection cannot provide guarantees because the database can fail after validation but before query execution.

If you *still* want to validate every connection, you can set the referenced system property to zero (0).
